crates/matrix-sdk/src/room_preview.rs

```diff
@@ -140,21 +144,76 @@ impl RoomPreview {
     ) -> crate::Result<Self> {
         // Use the room summary endpoint, if available, as described in
         // https://github.com/deepbluev7/matrix-doc/blob/room-summaries/proposals/3266-room-summary.md
-        match Self::from_room_summary(client, room_id.clone(), room_or_alias_id, via).await {
+        match Self::from_room_summary(client, room_id.clone(), room_or_alias_id, via.clone()).await
+        {
             Ok(res) => return Ok(res),
             Err(err) => {
                 warn!("error when previewing room from the room summary endpoint: {err}");
             }
         }
 
-        // TODO: (optimization) Use the room search directory, if available:
-        // - if the room directory visibility is public,
-        // - then use a public room filter set to this room id
+        // Try room directory search next.
+        match Self::from_room_directory_search(client, &room_id, room_or_alias_id, via).await {
+            Ok(res) => return Ok(res),
+            Err(err) => {
+                warn!("Room '{room_or_alias_id}' not found in room directory search: {err}");
+            }
+        }
 
         // Resort to using the room state endpoint, as well as the joined members one.
         Self::from_state_events(client, &room_id).await
     }
 
+    /// Get a [`RoomPreview`] by searching in the room directory for the
+    /// provided room alias or room id and transforming the [`RoomDescription`]
+    /// into a preview.
+    pub(crate) async fn from_room_directory_search(
+        client: &Client,
```

Comment:
nit: since the only thing done with this client is cloning it, let it be passed by ownership here (`Client`, not `&Client`), so the caller decides how to clone it instead. Doing this in general helps avoiding deep clones, in case the caller could just pass it by ownership.

Reply:
(still unresolved)

Reply:
Maybe I'm wrong, but:

- If we don't pass any via servers, we just clone it once (same as if we just passed the owned value).
- If we pass servers though, we need to clone it once per `RoomDirectorySearch::new(client)` call. If we already had an owned value, we'd be calling clone an extra time unless the do something weird like `let client = if last_iteration { client } else { client.clone() };`.

Reply:
Ah hm, we'd be cloning one extra if we have a `via` servers. Nevermind me, thanks for pushing back :)